lib/unary.ml

```diff
@@ -2,35 +2,111 @@ module type S = sig
   type _ t
 end
 
+module type SFold = sig
+  include S
+  type ('r,'s) cata
+  val map : ('a -> 'b) -> 'a t -> 'b t
+  val fold : ('r,'s) cata -> ('a -> 'r) -> 'a t -> 's
+end
+
+module ListF = struct 
+  type 'tl t = Cons of int * 'tl
+
+  type ('r,'rt) cata = {cata_cons:int -> 'r -> 'rt}
+
+  let map f (Cons(x,xs)) = Cons(x, f xs)
+
+  let fold : 'a 'r 'rt. ('r,'rt) cata -> ('a -> 'r) -> 'a t -> 'rt = 
+    fun cata self (Cons(x,xs)) ->
+    cata.cata_cons x (self xs)
+
+  let print (Cons(x,xs)) f = Printf.sprintf "Cons(%d, %s)" x (f xs)
+end
+let cata_ex =
+  {ListF.cata_cons = (fun x r -> x + r)}
+
 module Cy(M:S) = struct 
   open M
   (* cy in PHOAS form *)
   type _ cy = 
   | Cy : ('v -> ('v cy) t) -> 'v cy
-  | Cy2 : (('v * 'v -> 'v cy t) * ('v * 'v -> 'v cy t)) -> 'v cy
   | Var : 'v -> 'v cy
 end
 
-module Fold(C:S)(B:S) = struct
+module Cy2(M:S)(N:S) = struct
+  (* cy in PHOAS form *)
+  type (_,_) cy2 = 
+  | Cy2 : (('v * 'w -> ('v,'w) cy2 M.t) * ('v * 'w -> ('v,'w) cy2 N.t)) -> ('v,'w) cy2
+  | Var1 : 'v -> ('v,'w) cy2
+  | Var2 : 'w -> ('v,'w) cy2
+end
+
+module CyFold(C:SFold)(B:S) = struct
 
   module CyC = Cy(C)
   module CyB = Cy(B)
 
   type 'v cyc = 'v CyC.cy
   type 'v cyb = 'v CyB.cy
 
-  type 'cyb cata = {cataf:'cyc. 'cyc C.t -> ('cyc -> 'cyb) -> 'cyb B.t}
+  let rec fold : 'v. ('v cyb, 'v cyb B.t) C.cata -> 'v cyc -> 'v cyb = 
+    fun (type v) (cata : (v cyb, v cyb B.t) C.cata) (cy:v cyc) -> 
+      match cy with
+      | Cy f ->
+        CyB.Cy (fun bx -> C.fold cata (fold cata) (f bx))
+      | Var (v) -> Var(v)
+end
 
-  let rec fold : 'v . 'v cyb cata -> 'v cyc -> 'v cyb = 
-    fun (type v) (cata : v cyb cata) (cy:v cyc) -> 
+module Cy2Fold(C1:SFold)(C2:SFold)(B1:SFold)(B2:SFold) = struct
+
+  module CyC = Cy2(C1)(C2)
+  module CyB = Cy2(B1)(B2)
+  module CyC1 = Cy(C1)
+  module CyC2 = Cy(C2)
+  module CyB1 = Cy(B1)
+  module CyB2 = Cy(B2)
+  module Fold1 = CyFold(C1)(B1)
+  module Fold2 = CyFold(C2)(B2)
+
+  type ('v,'w) cyc = ('v,'w) CyC.cy2
+  type ('v,'w) cyb = ('v,'w) CyB.cy2
+  type 'v cyb1 = 'v CyB1.cy
+  type 'v cyb2 = 'v CyB2.cy
+
+  let rec bekic : 'v 'w. ('v,'w cyb2) cyb -> 'v cyb1 * 'w cyb2 =
+    let fst_ y = B1.map (fun x -> fst @@ bekic x) y in
+    let snd_ y = B2.map (fun x -> snd @@ bekic x) y in
+    fun (type v w) (cyb : (v,w cyb2) cyb) ->
+      match cyb with
+      | Cy2 (t,s) -> 
+        (CyB1.Cy (fun x -> fst_ @@ t (x, CyB2.Cy (fun y -> snd_ @@ s (x,y)))), 
```

Comment:
At `y` of `s (x,y)`:
> This expression has type w but an expression was expected of type
>   w cyb2

This error comes from the fact that Bekič rule 
```
cy(x,y. <T,S>) ==>
  <cy(x. (y.T) @@ cy(y.S))
   cy(y. (x.S) @@ cy(x. (y.T) @@ cy(y.S)))>
```
requires polymorphism at `y`. 
See that, in the first component `cy(x. (y.T) @@ cy(y.S))`,
the first `y` is replaced by `cy(y.S)`, which has type `w cyb2`, while
the second `y` must have type `w` by the very reason that `cy(y.S)` have type `w cyb2`.

Reply:
Looks like we need second-order polymorphism here.

Reply:
Ah, we don't need recursion here. Fixing